### PR TITLE
Refactor join.go

### PR DESCRIPTION
**Pull request by Jaanki**
Move non-join related functions out of join.go

Signed-ff-by: Janki Chhatbar <jchhatba@redhat.com>

### CODE REVIEW

File: pkg/subctl/cmd/join.go [hunk 1]
```diff
@@ -197,25 +197,12 @@ func joinSubmarinerCluster(config *rest.Config, subctlData *datafile.SubctlData)
...
-	exitOnError("Error deploying the operator", err)
-
-	if subctlData.ServiceDiscovery {
-		status.Start("Deploying multi cluster service discovery")
-		err = lighthouse.Ensure(status, config, "", "", false, kubeConfig, kubeContext)
-		status.End(err == nil)
-		exitOnError("Error deploying multi cluster service discovery", err)
-	}
-
+    var err error
```
Comment (dfarrell07) [inline pkg/subctl/cmd/join.go:205]: Seems like there's a minor formatting issue:

```
[00:01:55.963120] [submariner-operator]$ goimports -d tools.go cmd pkg
[00:01:56.110947] diff -u pkg/subctl/cmd/join.go.orig pkg/subctl/cmd/join.go
[00:01:56.111027] --- pkg/subctl/cmd/join.go.orig	2020-04-30 13:37:14.183365125 +0000
[00:01:56.111042] +++ pkg/subctl/cmd/join.go	2020-04-30 13:37:14.183365125 +0000
[00:01:56.111051] @@ -202,7 +202,7 @@
[00:01:56.111059]  		exitOnError("Unable to set the gateway node up", err)
[00:01:56.111068]  	}
[00:01:56.111077]  
[00:01:56.111085] -    var err error
[00:01:56.111093] +	var err error
```

Comment (Jaanki) [reply pkg/subctl/cmd/join.go]: DOne

File: pkg/discovery/globalnet/globalnet.go [hunk 2]
```diff
@@ -227,3 +246,143 @@ func nextPowerOf2(n uint32) uint {
...
+			return fmt.Errorf("unable to validate overlapping CIDR: %s", err)
+		}
+		if overlap && k != netconfig.ClusterId {
+			return fmt.Errorf("invalid CIDR: %s overlaps with cluster %s", cidr, k)
+		}
+	}
+	return nil
+}
+
+func CidrIsPreConfigured(clusterID string, globalNetworks map[string]*GlobalNetwork) bool {
```
Comment (tpantelis) [inline pkg/discovery/globalnet/globalnet.go:275]:
```suggestion
func isCIDRPreConfigured(clusterID string, globalNetworks map[string]*GlobalNetwork) bool {
```

File: pkg/discovery/globalnet/globalnet.go [hunk 3]
```diff
@@ -227,3 +246,143 @@ func nextPowerOf2(n uint32) uint {
...
+		if resource == 0 {
+			cidrlist = v.ClusterCIDRs
+			cidr = netconfig.ClusterCIDR
+		} else if resource == 1 {
+			cidrlist = v.ServiceCIDRs
+			cidr = netconfig.ServiceCIDR
+		} else if resource == 2 {
+			cidrlist = v.GlobalCIDRs
+			cidr = netconfig.GlobalnetCIDR
+		}
```
Comment (tpantelis) [inline pkg/discovery/globalnet/globalnet.go:263]: Use the type constants and a switch:

```
var cidrlist []string
var cidr string
switch overlappingFor {
case Cluster:
    cidrlist = v.ClusterCIDRs
    cidr = netconfig.ClusterCIDR
case Service:
    cidrlist = v.ServiceCIDRs
    cidr = netconfig.ServiceCIDR
case Global:
    cidrlist = v.GlobalCIDRs
    cidr = netconfig.GlobalnetCIDR
}

```

File: pkg/discovery/globalnet/globalnet.go [hunk 4]
```diff
@@ -47,7 +49,24 @@ type CIDR struct {
...
+	ClusterCIDR          string
+	ClusterId            string
+}
+
+type overlapCheck int
+
+const (
+	Cluster overlapCheck = iota
+	Service
+	Global
```
[the remark this answers is not in the document]
Comment (Jaanki) [reply pkg/discovery/globalnet/globalnet.go:65]: DOne

File: pkg/discovery/globalnet/globalnet.go [hunk 5]
```diff
@@ -47,7 +49,24 @@ type CIDR struct {
 	lastIp  uint
 }
 
+type Netconfig struct {
```
[the remark this answers is not in the document]
Comment (Jaanki) [reply pkg/discovery/globalnet/globalnet.go:52]: I think Netconfig is more clearer as it has configs for a particulat network.

File: pkg/discovery/globalnet/globalnet.go [hunk 6]
```diff
@@ -47,7 +49,24 @@ type CIDR struct {
...
+	GlobalnetClusterSize uint
+	GlobalnetCIDR        string
+	ServiceCIDR          string
+	ClusterCIDR          string
+	ClusterId            string
```
[the remark this answers is not in the document]
Comment (Jaanki) [reply pkg/discovery/globalnet/globalnet.go:57]: Done

File: pkg/discovery/globalnet/globalnet.go [hunk 7]
```diff
@@ -227,3 +246,143 @@ func nextPowerOf2(n uint32) uint {
 	n++
 	return uint(n)
 }
+
+func CheckOverlappingCidr(networks map[string]*GlobalNetwork, resource overlapCheck, netconfig Netconfig) error {
```
[the remark this answers is not in the document]
Comment (Jaanki) [reply pkg/discovery/globalnet/globalnet.go:250]: Done

File: pkg/discovery/globalnet/globalnet.go [hunk 8]
```diff
@@ -47,7 +49,24 @@ type CIDR struct {
...
 
+type Netconfig struct {
+	GlobalnetClusterSize uint
+	GlobalnetCIDR        string
+	ServiceCIDR          string
+	ClusterCIDR          string
+	ClusterId            string
+}
+
+type overlapCheck int
```
[the remark this answers is not in the document]
Comment (Jaanki) [reply pkg/discovery/globalnet/globalnet.go:60]: Done

File: pkg/discovery/globalnet/globalnet.go (hunk 2 above)
Comment (Jaanki) [reply pkg/discovery/globalnet/globalnet.go:275]: Done

File: pkg/discovery/globalnet/globalnet.go (hunk 3 above)
Comment (Jaanki) [reply pkg/discovery/globalnet/globalnet.go:263]: Its just about personal code style with such a small loop. I have anyways changed it :)

File: pkg/discovery/globalnet/globalnet.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (tpantelis) [reply pkg/discovery/globalnet/globalnet.go:52]: The package is `globalnet` so access reads `globalnet.Netconfig` so "net" seems redundant. Also this seems related to a cluster so to add more clarity I'd call it `ClusterConfig` or just `Config` I think is fine.

[the remark this answers is not in the document]
Comment (Jaanki) [reply pkg/discovery/globalnet/globalnet.go:52]: Done

**Comment by mangelajo** [comment]
Seems good to me, please wait for release tag before we merge this. Let's make it part of 0.4.0

**Comment by mangelajo** [comment]
This can go in now..

**Pull request merged**